### PR TITLE
Switch to `snafu` for errors types

### PR DESCRIPTION
**Pull request by danieldk**
`snafu` makes it easier to for us to have detailed errors by attaching context to errors. We still need to figure a bunch things, e.g.:

* I tried to make the errors as semantic as possible. So errors that bubble up are wrapped in semantic classes. E.g. in self-attention, we'd have error types to singal that something went wrong with: calculating attention scores, calculating temperature, weighting representation with the scores.

  However, it's not always clear how granular these error categories should be. Also, for some things it's not really clear where they belong, such as layer norm + dropout after summing all different types of embeddings.
* How to deal with the `Module`/`ModuleT` traits. They may not be great for our modules, since they require use of the candle error type.

The goal of this PR is put the error type/handling infrastructure in place, so that we can refine it in subsequent PRs.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -11,4 +11,5 @@ homepage = "https://github.com/danieldk/oxidized-transformers"
 [dependencies]
 candle-core = "0.3"
 candle-nn = "0.3"
+snafu = "0.8"
 thiserror = "1"
```
Comment (shadeMe) [inline Cargo.toml]: Do we still need the `thiserror` dependency?

File: src/layers/attention/mask.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    ApplyLogitsMask { source: candle_core::Error },
+
+    #[snafu(display("Cannot intersect masks"))]
+    IntersectMasks { source: candle_core::Error },
+}
+
+/// Attention mask.
+///
+/// Sequence elements for which the corresponding mask element is set to
+/// ``False`` are ignored during attention calculation.
```
Comment (shadeMe) [inline src/layers/attention/mask.rs]:
```suggestion
/// `False` are ignored during attention calculation.
```

File: src/layers/attention/mask.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#[derive(Clone, Debug)]
+pub struct AttentionMask {
+    bool_mask: Tensor,
+}
+
+impl AttentionMask {
+    /// Create an attention mask.
+    ///
+    /// * `bool_mask` - Boolean mask tensor.
+    ///   *Shape:* ``[batch_size, seq_len]``
```
Comment (shadeMe) [inline src/layers/attention/mask.rs]:
```suggestion
    ///   *Shape:* `(batch_size, seq_len)`
```

File: src/kv_cache.rs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+use candle_core::Tensor;
+
+pub struct KeyValueCache {
```
Comment (shadeMe) [inline src/kv_cache.rs:4]:
```suggestion
/// Cache type for layers that cache keys and values.
pub struct KeyValueCache {
```

File: src/layers/attention/mask.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    pub fn new(bool_mask: Tensor) -> Self {
+        AttentionMask { bool_mask }
+    }
+
+    /// Use the attention mask to mask attention logits.
```
Comment (shadeMe) [inline src/layers/attention/mask.rs:32]:
```suggestion
    /// Use the attention mask to mask attention logits.
    /// * input - Tensor to which the mask is applied.
    ///   *Shape:* `(batch_size, heads, query_len, key_len)`
    ///
    /// Returns: Logits with the attention mask applied.
	/// *Shape:* `(batch_size, heads, query_len, key_len)`
```

File: src/layers/attention/sdpa.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    AttentionMask { source: AttentionMaskError },
+
+    #[snafu(display("Cannot weigh representations using attention mask"))]
+    AttentionWeight { source: candle_core::Error },
+
+    #[snafu(display("Cannot apply softmax temperature"))]
+    Temperature { source: candle_core::Error },
+}
+
+pub struct ScaledDotProductAttention {
```
Comment (shadeMe) [inline src/layers/attention/sdpa.rs:35]:
```suggestion
/// Scaled dot-product attention.
///
/// See [Vaswani et al., 2017](https://arxiv.org/abs/1706.03762).
pub struct ScaledDotProductAttention {
```

File: src/layers/attention/sdpa.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    dropout: Dropout,
+    linear_biases: Option<AttentionLinearBiases>,
+}
+
+impl ScaledDotProductAttention {
+    pub fn new(dropout: Dropout, linear_biases: Option<AttentionLinearBiases>) -> Self {
```
Comment (shadeMe) [inline src/layers/attention/sdpa.rs:45]:
```suggestion
	/// Construct a scaled dot-product attention module.
	///
    /// * dropout - Dropout to apply to the final hidden representation.
    /// * linear_biases - ALiBi for attention scores. Not applied if `None`.
    pub fn new(dropout: Dropout, linear_biases: Option<AttentionLinearBiases>) -> Self {
```

File: src/layers/attention/self_attention.rs [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+use candle_core::{IndexOp, Module, ModuleT, Tensor};
+use candle_nn::{linear, Dropout, Linear, VarBuilder};
+use snafu::{ResultExt, Snafu};
+
+use crate::layers::attention::{
+    AttentionMask, AttentionMaskError, AttentionScorer, ScaledDotProductAttentionError,
+};
+use crate::layers::embeddings::{QueryKeyRotaryEmbeddings, QueryKeyRotaryEmbeddingsError};
+
+pub struct AttentionHeads {
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:14]:
```suggestion
/// Attention heads configuration to use in self-attention.
pub struct AttentionHeads {
```

File: src/layers/attention/self_attention.rs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    n_query_heads: usize,
+    n_key_value_heads: usize,
+    qkv_mode: QkvMode,
+}
+
+#[non_exhaustive]
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:26]:
```suggestion
/// Query, key, value splitting strategies.
///
/// After the input projection of the attention layer, we have an array with
/// shape `(batch_size, seq_len, n_heads * head_width)` where `n_heads` is
/// the sum of the number of query, key, and value heads. We need to split up
/// the array into separate arrays for query, key, and value heads.
#[non_exhaustive]
```

File: src/layers/attention/self_attention.rs [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+pub enum QkvSplit {
+    Default,
+    KVSizedChunks,
+}
+
+#[non_exhaustive]
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:34]:
```suggestion
/// How the query, key and value projections are handled in
/// the self-attention layer.
#[non_exhaustive]
```

File: src/layers/attention/self_attention.rs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+pub enum QkvMode {
+    Separate,
+    MergedSplitBefore,
+    MergedSplitAfter(QkvSplit),
+}
+
+pub enum QkvTensors {
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:42]:
```suggestion
/// Representation of the query, key and value tensors.
pub enum QkvTensors {
```

File: src/layers/attention/mod.rs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    /// Returns: Attention values.
+    /// *Shape:* `(batch_size, heads, seq_len, width)`
+    fn forward(
+        &self,
+        query: &Tensor,
+        key: &Tensor,
+        value: &Tensor,
+        attention_mask: &AttentionMask,
+        train: bool,
+    ) -> Result<Tensor, Box<dyn Error + Send + Sync>>;
```
Comment (shadeMe) [inline src/layers/attention/mod.rs:42]: I guess there's no way around boxed error representations, right?

File: src/layers/attention/self_attention.rs [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    },
+
+    #[snafu(display("Cannot construct layer"))]
+    SelfAttentionConstruction { source: candle_core::Error },
+
+    #[snafu(display("Cannot split heads"))]
+    SplitHeads { source: candle_core::Error },
+}
+
+pub struct SelfAttention {
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:97]:
```suggestion
/// Transformer self-attention layer.
///
/// See [Vaswani et al., 2017](https://arxiv.org/abs/1706.03762).
pub struct SelfAttention {
```

File: src/layers/embeddings/qk_rotary_embeddings.rs [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+pub struct QueryKeyRotaryEmbeddings {
+    rotary_width: usize,
+    rotary_embeds: RotaryEmbeddings,
+    pub head_width: usize,
+}
+
+impl QueryKeyRotaryEmbeddings {
+    /// Construct query-key rotary embeddings module.
+    ///
+    /// The rotary embedding will be precomputed for up to ``seq_len`` positions.
```
Comment (shadeMe) [inline src/layers/embeddings/qk_rotary_embeddings.rs]:
```suggestion
    /// The rotary embedding will be precomputed for up to `seq_len` positions.
```

File: src/layers/embeddings/qk_rotary_embeddings.rs [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    /// The embedding will be recomputed when a longer sequence is found in the
+    /// input.
+    ///
+    /// * `fraction` - Fraction of hidden width to apply rotary embeddings to.
+    ///   Must be in ``[0,1]``.
```
Comment (shadeMe) [inline src/layers/embeddings/qk_rotary_embeddings.rs]:
```suggestion
    ///   Must be in `[0,1]`.
```

File: src/layers/embeddings/qk_rotary_embeddings.rs [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+    /// * `head_width` - Query and key attention head width.
+    /// * `seq_len` - Number of positions to initially precompute.
+    /// * `base` - The base used for theta_i (normally 10_000). Determines the cycle
```
Comment (shadeMe) [inline src/layers/embeddings/qk_rotary_embeddings.rs]:
```suggestion
    /// * `base` - The base used for `theta` (normally 10_000). Determines the cycle
```
Typo?

File: src/layers/embeddings/qk_rotary_embeddings.rs [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+    /// Returns the input with the rotary embeddings applied. *Shape:*
+    /// `(batch_size, n_heads, seq_len, width_per_head)`
+    ///
+    /// * `query` - Query to apply the rotary embeddings to.
+    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
+    /// * `key` - Key to apply the rotary embeddings to.
+    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
+    /// * `positions` - Positions of the inputs. If no positions are
+    ///    provided, they are assumed to be `[0, seq_len)`.
+    ///    *Shape:* `(batch_size, seq_len)`
```
Comment (shadeMe) [inline src/layers/embeddings/qk_rotary_embeddings.rs]:
```suggestion
    /// * `query` - Query to apply the rotary embeddings to.
    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
    /// * `key` - Key to apply the rotary embeddings to.
    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
    /// * `positions` - Positions of the inputs. If no positions are
    ///    provided, they are assumed to be `[0, seq_len)`.
    ///    *Shape:* `(batch_size, seq_len)`
    ///
    /// Returns: Input with the rotary embeddings applied. 
    /// *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
```

File: src/layers/embeddings/qk_rotary_embeddings.rs [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    pub fn forward(
+        &self,
+        query: &Tensor,
+        key: &Tensor,
+        cache: Option<KeyValueCache>,
+        mut positions: Option<Tensor>,
+    ) -> Result<(Tensor, Tensor), QueryKeyRotaryEmbeddingsError> {
+        let (batch_size, n_heads, seq_len, head_width) =
+            query.shape().dims4().context(InvalidRankSnafu {
+                expected: 4usize,
```
Comment (shadeMe) [inline src/layers/embeddings/qk_rotary_embeddings.rs:102]:
```suggestion
                expected: 4,
```
Can be inferred/omitted here, no?

File: src/layers/embeddings/rotary_embeddings.rs [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+#[derive(Debug)]
+pub struct RotaryEmbeddings {
+    cache: RwLock<RotaryEmbeddingsCache>,
+    theta: Tensor,
+}
+
+impl RotaryEmbeddings {
+    /// Construct a rotary embedding module.
+    ///
+    /// The rotary embedding will be precomputed for up to ``seq_len`` positions.
```
Comment (shadeMe) [inline src/layers/embeddings/rotary_embeddings.rs:63]:
```suggestion
    /// The rotary embedding will be precomputed for up to `seq_len` positions.
```

File: src/layers/embeddings/rotary_embeddings.rs [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    /// The embedding will be recomputed when a longer sequence is found in the
+    /// input.
+    ///
+    /// * `width` - Rotary embedding width. Must be even.
+    /// * `seq_len` - Number of positions to initially precompute.
+    /// * `base` - The base used for theta_i (normally 10_000). Determines the cycle
```
Comment (shadeMe) [inline src/layers/embeddings/rotary_embeddings.rs]:
```suggestion
    /// * `base` - The base used for `theta` (normally 10_000). Determines the cycle
```

File: src/layers/embeddings/rotary_embeddings.rs [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+    /// Apply rotary embeddings to the input.
+    ///
+    /// Returns the input with the rotary embeddings applied. *Shape:*
+    /// `(batch_size, n_heads, seq_len, width_per_head)`
+    ///
+    /// * `input` - Input to apply the rotary embeddings to.
+    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
+    /// * `positions` - Positions of the inputs. If no positions are
+    ///    provided, they are assumed to be `[0, seq_len)`.
+    ///    *Shape:* `(batch_size, seq_len)`
```
Comment (shadeMe) [inline src/layers/embeddings/rotary_embeddings.rs]:
```suggestion
    /// * `input` - Input to apply the rotary embeddings to.
    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
    /// * `positions` - Positions of the inputs. If no positions are
    ///    provided, they are assumed to be `[0, seq_len)`.
    ///    *Shape:* `(batch_size, seq_len)`
    ///
    /// Returns: Input with the rotary embeddings applied. 
    /// *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`    
```

File: src/layers/feedforward.rs [hunk 22]
```diff
@@ -29,29 +27,36 @@ use crate::error::Result;
...
+    layer_norm: Box<dyn ModuleT>,
+    output: Linear,
 }
 
 impl PointwiseFeedForward {
     /// Construct a point-wise feed-forward layer.
     ///
     /// * `vb` - Variable store.
     /// * `activation` - Non-linearity.
+    /// * `dropout` - Probability of the dropout applied to the output of the layer.
```
Comment (shadeMe) [inline src/layers/feedforward.rs]:
```suggestion
    /// * `dropout` - Dropout applied to the output of the layer.
```

File: src/layers/feedforward.rs [hunk 23]
```diff
@@ -29,29 +27,36 @@ use crate::error::Result;
...
     /// * `hidden_width` - Hidden width, dimensionality of the layer input and output.
     /// * `intermediate_width` - Intermediate width inside the feed-forward layer.
+    /// * `layer_norm` - Layer normalization to apply to the input.
```
Comment (shadeMe) [inline src/layers/feedforward.rs]:
```suggestion
    /// * `layer_norm` - Layer normalization applied to the input.
```

File: src/layers/feedforward.rs [hunk 24]
```diff
@@ -1,7 +1,5 @@
-use candle_core::{Module, Tensor};
-use candle_nn::{linear, linear_no_bias, Linear, VarBuilder};
-
-use crate::error::Result;
+use candle_core::{Module, ModuleT, Tensor};
+use candle_nn::{linear, linear_no_bias, Dropout, Linear, VarBuilder};
```
Comment (shadeMe) [inline src/layers/feedforward.rs:2]: No errors for this module?

**Review by shadeMe — APPROVED** [review]
LGTM 🎉 Added a bunch of documentation suggestions and couple of questions. Good to merge after applying the suggestions.

File: src/layers/feedforward.rs (hunk 24 above)
Comment (danieldk) [reply src/layers/feedforward.rs:2]: The issue is that this module implements `ModuleT` and we cannot implement that with a custom error type. I think in a follow-up PR, we should define our own module trait that is more flexible.

File: src/layers/embeddings/qk_rotary_embeddings.rs (hunk 18 above)
Comment (danieldk) [reply src/layers/embeddings/qk_rotary_embeddings.rs:102]: Oddly enough type inference fails in this case.

File: src/layers/attention/mod.rs (hunk 12 above)
Comment (danieldk) [reply src/layers/attention/mod.rs:42]: We could make the error an associated type for `AttentionScorer`, but then we have to specify the error type in `SelfAttention` somehow, which is probably not nice either?.

File: Cargo.toml (hunk 1 above)
Comment (danieldk) [reply Cargo.toml]: Fixed.

File: src/layers/attention/mod.rs (hunk 12 above)
Comment (shadeMe) [reply src/layers/attention/mod.rs:42]: Yeah, agreed.
